aten/src/ATen/native/xnnpack/Convolution.cpp

```diff
@@ -0,0 +1,302 @@
+      stride[Layout::Parameter::Width],                               // subsampling_width
+      dilation[Layout::Parameter::Height],                            // dilation_height
+      dilation[Layout::Parameter::Width],                             // dilation_width
+      groups,                                                         // groups
+      weight_nhwc.size(Layout::Filter::Input),                        // group_input_channels
+      weight_nhwc.size(Layout::Filter::Output) / groups,              // group_output_channels
+      weight_nhwc.size(Layout::Filter::Input) * groups,               // input_pixel_stride
+      weight_nhwc.size(Layout::Filter::Output),                       // output_pixel_stride
+      weight_nhwc.data_ptr<float>(),                                  // kernel
+      (bias && bias->defined()) ? bias->data_ptr<float>() : nullptr,  // bias
+      output_min.to<float>(),                                         // output_min
+      output_max.to<float>(),                                         // output_max
+      0u,                                                             // flags
+      &convolution_op);                                               // operator
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == create_status,
+      "xnn_create_convolution2d_nhwc_f32 failed!");
+
+  const auto Spatial = [&](const size_t filter, const size_t parameter) {
+    const int64_t kernel = 1 + dilation[parameter] * (weight_nhwc.size(filter) - 1);
+
+    return Context::Output::Spatial{
+      stride[parameter] - kernel + 2 * padding[parameter],
+      stride[parameter],
+    };
+  };
+
+  return Context{
+    Operator(convolution_op),
+    {
+      weight_nhwc.size(Layout::Filter::Output),
+      Spatial(Layout::Filter::Height, Layout::Parameter::Height),
+      Spatial(Layout::Filter::Width, Layout::Parameter::Width),
+    },
+  };
+}
+
+bool usable(const Tensor& input) {
+         // Input
+  return (4 == input.ndimension()) &&
+         (c10::DeviceType::CPU == input.device().type()) &&
+         (kFloat == input.scalar_type()) &&
+         (input.size(Layout::Activation4D::Batch) > 0) &&
+         (input.size(Layout::Activation4D::Channels) > 0) &&
+         (input.size(Layout::Activation4D::Height) > 0) &&
+         (input.size(Layout::Activation4D::Width) > 0);
+}
+
+Tensor run(
+    const Context& context,
+    const Tensor& input) {
+  using namespace internal;
+
+  const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
+
+  TORCH_CHECK(
+      usable(input_nhwc),
+      "xnnpack::convolution not usable!");
+
+  const auto Spatial = [](const int64_t input, const Context::Output::Spatial spatial) {
+    return (input + spatial.alpha) / spatial.beta;
+  };
+
+  Tensor output = empty(
+      std::array<int64_t, 4>{
+        input_nhwc.size(Layout::Activation4D::Batch),
+        context.output.channels,
```

Comment:
Shouldn't the order of dims here be NHWC? or you always provided dimension sizes in NCHW and then convert to channel last format?

Reply:
No, unless I am mistaken, the dimensions are always specified in NCHW format even if your tensor is in NHWC format.  That is by design.

Reply:
Ashkan, if this is the case then for `ActivationND::Channel` should return dimensions accordingly, no? `ActivationND::Channel`, for example, returns the last dimension. Does the comment you have above, apply only to the empty call?